db/src/tx.rs

```diff
@@ -245,157 +348,133 @@ impl<'conn, 'a> Tx<'conn, 'a> {
                 },
 
                 Entity::AddOrRetract { op, e, a, v } => {
-                    let a: i64 = match a {
-                        entmod::Entid::Entid(ref a) => *a,
-                        entmod::Entid::Ident(ref a) => self.schema.require_entid(&a)?,
-                    };
-
-                    let attribute: &Attribute = self.schema.require_attribute_for_entid(a)?;
-
-                    let v = match v {
-                        entmod::AtomOrLookupRefOrVectorOrMapNotation::Atom(v) => {
-                            if attribute.value_type == ValueType::Ref && v.inner.is_text() {
-                                Either::Right(LookupRefOrTempId::TempId(temp_ids.intern(v.inner.as_text().cloned().map(TempId::External).unwrap())))
-                            } else {
-                                // Here is where we do schema-aware typechecking: we either assert that
-                                // the given value is in the attribute's value set, or (in limited
-                                // cases) coerce the value into the attribute's value set.
-                                let typed_value: TypedValue = self.schema.to_typed_value(&v.without_spans(), &attribute)?;
-                                Either::Left(typed_value)
-                            }
-                        },
-
-                        entmod::AtomOrLookupRefOrVectorOrMapNotation::LookupRef(lookup_ref) => {
-                            if attribute.value_type != ValueType::Ref {
-                                bail!(ErrorKind::NotYetImplemented(format!("Cannot resolve value lookup ref for attribute {} that is not :db/valueType :db.type/ref", a)))
-                            }
-
-                            Either::Right(LookupRefOrTempId::LookupRef(intern_lookup_ref(&mut lookup_refs, lookup_ref)?))
-                        },
+                    if a.is_backward() {
+                        let reversed_e = in_process.entity_v_into_term_e(v, &a)?;
+                        let (reversed_a, _reversed_attribute) = in_process.entity_a_into_term_a(a.to_reversed().unwrap())?;
```

Comment:
Same point here about reversing. I think if you switch these two `is_backward`/`to_reversed` pairs, you can get rid of `to_reversed`.